### ISSUE TITLE
Inconsistent results from noisy density matrix simulator

### ISSUE THREAD

**Issue by viathor** [quantumlib/Cirq#4163]
**Description of the issue**

There are two ways of getting noise into a density matrix simulation: we can attach the noise to the simulator (by passing the `noise` parameter to the constructor) or to the circuit (by calling `with_noise`). I think it is reasonable to expect the two to produce the same results. However, I found that they don't always agree.

I tried to create a minimal command to reproduce this in ipython and noticed that the final density matrix is unreliable. Most of the time it looks OK, but occasionally it is wrong. This afflicts both ways of attaching noise.

**How to reproduce the issue**

See unit test added in #4162. The test passes most of the time. Sometimes it fails. Here is example failure from CI (it fails on my machine the same way):

```
_______________ test_noisy_sim_vs_noisy_circuit[noise1-circuit1] _______________

circuit = cirq.Circuit([
    cirq.Moment(
        cirq.X(cirq.LineQubit(0)),
    ),
    cirq.Moment(
        cirq.measure(cirq.LineQubit(0)),
    ),
])
noise = cirq.amplitude_damp(gamma=0.1)

    @pytest.mark.parametrize(
        'circuit',
        (
            cirq.Circuit(cirq.X(q0)),
            cirq.Circuit(cirq.X(q0), cirq.measure(q0)),
        ),
    )
    @pytest.mark.parametrize('noise', (cirq.depolarize(0.1), cirq.amplitude_damp(0.1)))
    def test_noisy_sim_vs_noisy_circuit(circuit, noise):
        rho1 = cirq.DensityMatrixSimulator(noise=noise).simulate(circuit).final_density_matrix
        rho2 = cirq.DensityMatrixSimulator().simulate(circuit.with_noise(noise)).final_density_matrix
        print(rho1)
        print(rho2)
>       assert np.allclose(rho1, rho2)
E       assert False
E        +  where False = <function allclose at 0x7fe41cdab680>(array([[1.+0.j, 0.+0.j],\n       [0.+0.j, 0.+0.j]], dtype=complex64), array([[0.1       +0.j, 0.        +0.j],\n       [0.        +0.j, 0.90000004+0.j]], dtype=complex64))
E        +    where <function allclose at 0x7fe41cdab680> = np.allclose

cirq-core/cirq/sim/density_matrix_simulator_test.py:1393: AssertionError
----------------------------- Captured stdout call -----------------------------
[[1.+0.j 0.+0.j]
 [0.+0.j 0.+0.j]]
[[0.1       +0.j 0.        +0.j]
 [0.        +0.j 0.90000004+0.j]]
=========================== short test summary info ============================
FAILED cirq-core/cirq/sim/density_matrix_simulator_test.py::test_noisy_sim_vs_noisy_circuit[noise1-circuit1]
======================== 1 failed, 158 passed in 1.49s =========================
```

Also, here is a log from ipython session that shows the unreliable results:

```
$ ipython
Python 3.9.2 (default, Feb 28 2021, 17:03:44) 
Type 'copyright', 'credits' or 'license' for more information
IPython 7.23.1 -- An enhanced Interactive Python. Type '?' for help.

In [1]: import cirq

In [2]: q = cirq.NamedQubit('q')

In [3]: noise = cirq.amplitude_damp(0.1)

[...]

In [7]: circuit = cirq.Circuit(cirq.H(q), cirq.measure(q))

In [8]: cirq.DensityMatrixSimulator(noise=noise).simulate(
   ...:     circuit
   ...: ).final_density_matrix
Out[8]: 
array([[0.9999999+0.j, 0.       +0.j],
       [0.       +0.j, 0.       +0.j]], dtype=complex64)

In [9]: cirq.DensityMatrixSimulator().simulate(
   ...:     circuit.with_noise(noise)
   ...: ).final_density_matrix
Out[9]: 
array([[0.09999999+0.j, 0.        +0.j],
       [0.        +0.j, 0.9       +0.j]], dtype=complex64)

In [10]: circuit = cirq.Circuit(cirq.X(q), cirq.measure(q))

In [11]: cirq.DensityMatrixSimulator(noise=noise).simulate(
    ...:     circuit
    ...: ).final_density_matrix
Out[11]: 
array([[0.1       +0.j, 0.        +0.j],
       [0.        +0.j, 0.90000004+0.j]], dtype=complex64)

In [12]: cirq.DensityMatrixSimulator().simulate(
    ...:     circuit.with_noise(noise)
    ...: ).final_density_matrix
Out[12]: 
array([[0.1       +0.j, 0.        +0.j],
       [0.        +0.j, 0.90000004+0.j]], dtype=complex64)

In [13]: circuit
Out[13]: q: ───X───M───

In [14]: cirq.DensityMatrixSimulator().simulate(
    ...:     circuit.with_noise(noise)
    ...: ).final_density_matrix
Out[14]: 
array([[1.+0.j, 0.+0.j],
       [0.+0.j, 0.+0.j]], dtype=complex64)

In [15]: 

In [15]:                                                                                                                                                                       
Do you really want to exit ([y]/n)? y
```

**Cirq version**
You can get the cirq version by printing `cirq.__version__`. From the command line:

```
$ python -c 'import cirq; print(cirq.__version__)'
0.12.0.dev
```

**Comment by viathor** [comment]
Actually, I guess what happens is the sim collapses the wavefunction different ways. The density matrices are consistent with that (since it appears that the amplitude damping channel is also added to measurements).

**Comment by balopat** [comment]
Maybe a little tangential as a remark but the effect of measurement itself could be captured as a sum of the effect of the projectors on the density matrix, i.e. capturing the measurement outcome in the "classical probability layer" of the density matrix, instead of actually simulating the probability pathway and collapsing the wavefunction. Was that your expectation maybe? Would that simulation mode be useful to you?

**Comment by viathor** [comment]
I didn't realize that noise channel was added after my measurement gate and so after seeing fractional entries in the output density matrix I concluded that it is not a collapsed state and thought that `final_density_matrix` was instead a pre-measurement state. The behavior made sense once I looked at the modified circuit.

For my purposes getting a collapsed wavefunction is sufficient, but I imagine that having a way to get a classical distribution over outcomes instead might be useful. Then again, it's ok to wait for the use-case to emerge before we add such a simulation mode.
